python/lsst/meas/base/_id_generator.py

```diff
@@ -0,0 +1,548 @@
+
+class FullIdGenerator(IdGenerator):
+    """The subclass of `IdGenerator` that actually includes packed data IDs
+    and release IDs in its generated IDs.
+
+    Parameters
+    ----------
+    dimension_packer : `lsst.daf.butler.DimensionPacker`
+        Object that packs data IDs into integers.
+    data_id : `lsst.daf.butler.DataCoordinate`
+        Data ID to embed in all generated IDs and random seeds.
+    release_id : `int`, optional
+        Release identifier to embed in generated IDs.
+    n_releases : `int`, optional
+        Number of (contiguous, starting from zero) `release_id` values to
+        reserve space for. One (not zero) is used to reserve no space.
+
+    Notes
+    -----
+    Instances of this class should usually be constructed via configuration
+    instead of by calling the constructor directly; see `IdGenerator` for
+    details.
+    """
+
+    def __init__(
+        self,
+        dimension_packer: DimensionPacker,
+        data_id: DataCoordinate,
+        release_id: int = DEFAULT_RELEASE_ID,
+        n_releases: int = DEFAULT_N_RELEASES,
+    ):
+        self._bits = _IdGeneratorBits(dimension_packer, n_releases)
+        self._release_id = release_id
+        self._data_id = data_id.subset(self._bits.packer.dimensions)
+        self._packed_data_id = self._bits.packer.pack(self._data_id)
+
+    @property
+    def data_id(self) -> DataCoordinate:
+        """The data ID that will be embedded in all generated IDs
+        (`DataCoordinate`)."""
+        return self._data_id
+
+    @property
+    def release_id(self) -> int:
+        """The release ID that will embedded in all generated IDs (`int`)."""
+
+    @property
+    def catalog_id(self) -> int:
+        # Docstring inherited.
+        return self._packed_data_id + self._bits.n_data_ids * self._release_id
+
+    def __str__(self) -> str:
+        # Docstring inherited.
+        return str(self.data_id)
+
+    def make_table_id_factory(self) -> IdFactory:
+        # Docstring inherited.
+        return IdFactory.makeSource(self.catalog_id, self._bits.counter_bits)
+
+    def arange(self, *args, **kwargs) -> np.ndarray:
+        # Docstring inherited.
+        lower = super().arange(*args, **kwargs)
+        if np.any(lower >= self._bits.n_counters):
+            arg_terms = [repr(arg) for arg in args] + [
+                f"{k}={v!r}" for k, v in kwargs.items()
+            ]
+            raise ValueError(
+                f"Integer range from numpy.arange({arg_terms}) has values that are not "
+                f"below upper bound of {self._bits.n_counters}."
+            )
+        return lower + self.catalog_id * self._bits.n_counters
+
+
+@dataclasses.dataclass
+class _IdGeneratorBits:
+    """A private helper struct that manages the allocation of bits between the
+    packed data ID, the release ID, and a per-catalog counter.
+    """
+
+    packer: DimensionPacker
+    """Object that maps data IDs to integers
+    (`lsst.daf.butler.DimensionPacker`).
```

Comment:
I'm a bit surprised that we have to duplicate types for dataclasses. Isn't that one of the points of the dataclass definition?

Reply:
It's just like any other type annotation as far as Sphinx is concerned right now.  I cannot way for the promised documenteer upgrade that will fix this.